reasoner/resolution/resolver/BoundConclusionResolver.java

```diff
@@ -0,0 +1,377 @@
+
+        ConclusionRequestState<?> requestState;
+        if (fromUpstream.partialAnswer().asConclusion().isExplain()) {
+            requestState = new ConclusionRequestState.Explain(fromUpstream, iteration, conditionDownstreams(fromUpstream));
+        } else if (fromUpstream.partialAnswer().asConclusion().isMatch()) {
+            requestState = new ConclusionRequestState.Match(fromUpstream, iteration, conditionDownstreams(fromUpstream));
+        } else {
+            throw TypeDBException.of(ILLEGAL_STATE);
+        }
+
+        return requestState;
+    }
+
+    private List<Request> conditionDownstreams(Request fromUpstream) {
+        // TODO: Can there be more than one downstream Condition? If not reduce this return type from List to single
+        // we do a extra traversal to expand the partial answer if we already have the concept that is meant to be generated
+        // and if there's extra variables to be populated
+        AnswerState.Partial.Conclusion<?, ?> partialAnswer = fromUpstream.partialAnswer().asConclusion();
+        assert fromUpstream.partialAnswer().isConclusion();
+        assert conclusion.retrievableIds().containsAll(partialAnswer.conceptMap().concepts().keySet());
+
+        List<Request> downstreams = new ArrayList<>();
+        if (conclusion.generating().isPresent() && conclusion.retrievableIds().size() > partialAnswer.conceptMap().concepts().size() &&
+                partialAnswer.conceptMap().concepts().containsKey(conclusion.generating().get().id())) {
+            candidateAnswers(partialAnswer).forEachRemaining(answer -> downstreams.add(Request.create(driver(), conditionResolver, answer)));
+        } else {
+            Set<Identifier.Variable.Retrievable> named = iterate(conclusion.retrievableIds()).filter(Identifier::isName).toSet();
+            downstreams.add(Request.create(driver(), conditionResolver, partialAnswer.toDownstream(named)));
+        }
+        return downstreams;
+    }
+
+    private FunctionalIterator<AnswerState.Partial.Compound<?, ?>> candidateAnswers(AnswerState.Partial.Conclusion<?, ?> partialAnswer) {
+        GraphTraversal traversal = boundTraversal(conclusion.conjunction().traversal(), partialAnswer.conceptMap());
+        FunctionalIterator<ConceptMap> answers = traversalEngine.iterator(traversal).map(conceptMgr::conceptMap);
+        Set<Identifier.Variable.Retrievable> named = iterate(conclusion.retrievableIds()).filter(Identifier::isName).toSet();
+        return answers.map(ans -> partialAnswer.extend(ans).toDownstream(named));
+    }
+
+    private static abstract class ConclusionRequestState<CONCLUDABLE extends Concludable<?>> extends RequestState {
+
+        private final DownstreamManager downstreamManager;
+        private final WaitedMaterialisations waitedMaterialisations;
+        private boolean complete;
+        protected FunctionalIterator<CONCLUDABLE> materialisations;
```

Comment:
should this be a `Poller`? since you extend and query it multiple times...

Reply:
The implementation has already been revamped using a `Deque`